doc/reference/reference_lua/config/utils_schema.rst

```diff
@@ -0,0 +1,1163 @@
+    :dedent:
+
+To create an array node in a schema, use :ref:`schema.array() <config-utils-schema-array>`.
+The following schema describes this configuration:
+
+..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_nodes_array/http-api.lua
+    :language: lua
+    :start-at: local listen_address_schema
+    :end-before: local function validate
+    :dedent:
+
+There is also the :ref:`schema.set() <config-utils-schema-set>` function that enables
+creating arrays with a limited set of allowed items.
+
+.. _config_utils_schema_nodes_map:
+
+Maps
+****
+
+*Map* is a composite node type that holds an arbitrary number of key-value pairs
+of predefined types.
+
+In YAML, a map is represented as a node with nested fields.
+For example, the following configuration has the ``endpoints`` node:
+
+..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_nodes_map/config.yaml
+    :language: yaml
+    :start-at: roles
+    :dedent:
+
+To create a map node in a schema, use :ref:`schema.map() <config-utils-schema-map>`.
+If this node is declared as a map as shown below, the ``endpoints`` section can include
+any number of options with arbitrary names and boolean values.
+
+..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_nodes_map/http_api.lua
+    :language: lua
+    :start-at: local listen_address_schema
+    :end-before: local function validate
+    :dedent:
+
+.. _config_utils_schema_annotation:
+
+Annotations
+***********
+
+Node *annotations* are named attributes that define the its various aspects. For example,
+scalar nodes have a required annotation ``type`` that defines the node value type.
+Other annotations can optionally set default values, validation function, and
+
+Annotations are passed in a table to the node creation function:
+
+..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_annotations/http_api.lua
+    :language: lua
+    :start-at: scheme = schema.scalar
+    :end-before: host = schema.scalar
+    :dedent:
+
+Node annotations fall into three groups:
+
+-   *Built-in annotations* are handled by the module. These are: ``type``, ``validate``, ``allowed_values``, ``default`` and ``apply_default_if``.
+    Note that ``validate``, ``allowed_values`` are used for validation only. ``default`` and ``apply_default_if`` can transform the configuration.
+-   *User-defined annotations* add named node attributes that can be used in the
+    application or role code.
+-   *Computed annotations* allow access to annotations of other nodes throughout
+    the schema.
+
+.. _config_utils_schema_built_in_annotations:
+
+Built-in annotations
+^^^^^^^^^^^^^^^^^^^^
+
+Built-in annotations are interpreted by the module itself. There are the following
+built-in annotations:
+
+-   :ref:`type <config-schema_node_annotation-type>` -- the node value type. Mandatory for scalar nodes, except for those created with ``schema.enum()``.
```

Comment:
> Mandatory for scalar nodes

The `type` field is mandatory for all the schema nodes (including ones of a composite type). However, there are schema node constructors that set it on its own: `schema.record()`, `schema.map()`, `schema.array()`, `schema.enum()`, `schema.set()`.

Let's describe this nuance in regards to the schema node *constructors*, not schema nodes.

Reply:
Added this info.